File: Pythoncode/63.不同路径 II.py
class Solution:
    def countPaths(self, index2NumofPaths, m, n, x, y):
        if (x, y) in index2NumofPaths.keys():
            return index2NumofPaths[(x, y)]
        if x == 0 and y == 0:
            index2NumofPaths[(x, y)] = 1
            return 1
        elif x < 0 or x >= m or y < 0 or y >= n:
            index2NumofPaths[(x, y)] = 0
            return 0
        else:
            index2NumofPaths[(x, y)] = self.countPaths(index2NumofPaths, m, n, x-1, y) + self.countPaths(index2NumofPaths, m, n, x, y-1)
            return index2NumofPaths[(x, y)]

    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        index2NumofPaths = {}
        n = len(obstacleGrid)
        m = len(obstacleGrid[0])
        for x in range(m):
            for y in range(n):
                if obstacleGrid[y][x] == 1:
                    index2NumofPaths[(x, y)] = 0
        return self.countPaths(index2NumofPaths, m, n, m-1, n-1)

File: Cppcode/63.不同路径II.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        auto m = obstacleGrid.size();
        auto n = obstacleGrid[0].size();
        vector<vector<int>> dp(m, vector<int>(n));
        for (auto i = 0; i < m; i++)
        {
            for (auto j = 0; j < n; j++)
            {
                if (obstacleGrid[i][j] == 1)
                {
                    dp[i][j] = 0;
                    continue;
                }
                if (i == 0 && j == 0)
                {
                    dp[i][j] = 1;
                    continue;
                }
                auto a = i-1>=0 ? dp[i-1][j] : 0;
                auto b = j-1>=0 ? dp[i][j-1] : 0;
                dp[i][j] = a + b;
            }
        }
        return dp[m-1][n-1];
    }
};